client/app.go

```diff
@@ -32,12 +33,43 @@ func (c *ThreeScaleClient) CreateApp(accessToken string, accountId string, planI
 
 	resp, err := c.httpClient.Do(req)
 	if err != nil {
-		return apiResp, genRespErr("create application", err.Error())
+		return apiResp, err
 	}
 	defer resp.Body.Close()
 
-	if err := xml.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
-		return apiResp, genRespErr("create application", err.Error())
+	err = handleResp(resp, http.StatusCreated, &apiResp)
+	return apiResp, err
+}
+
+func handleResp(resp *http.Response, expectCode int, decodeInto interface{}) error {
+	err := handleErrResp2(resp, expectCode)
+	if err != nil {
+		return err
+	}
+
+	if err := xml.NewDecoder(resp.Body).Decode(decodeInto); err != nil {
+		msg := fmt.Sprintf("error decoding api response - %s", err.Error())
+		return ApiErr{
+			code: resp.StatusCode,
+			err:  msg,
+		}
+	}
+	return nil
+}
+
+// Decodes and transforms an API response error into a string
+func handleErrResp2(resp *http.Response, expectCode int) error {
+	if resp.StatusCode != expectCode {
```

Comment:
this func does nothing when the status code is the expected one, so how about only calling it when we know for sure an error will be returned?

Reply:
Yep, makes sense